File: demo/simple search/YahtzeeStatus.py
class ScoreType(int):
    Ones = 1 << 0
    Twos = 1 << 1
    Threes = 1 << 2
    Fours = 1 << 3
    Fives = 1 << 4
    Sixes = 1 << 5
    Chance = 1 << 6
    ThreeOfAKind = 1 << 7
    FourOfAKind = 1 << 8
    FullHouse = 1 << 9
    SmallStraight = 1 << 10
    LargeStraight = 1 << 11
    Yahtzee = 1 << 12
    Bonus = 1 << 13


class YahtzeeStatus():
    ALL_SCORE_TYPE = [ScoreType.Ones,
                      ScoreType.Twos,
                      ScoreType.Threes,
                      ScoreType.Fours,
                      ScoreType.Fives,
                      ScoreType.Sixes,
                      ScoreType.Chance,
                      ScoreType.FourOfAKind,
                      ScoreType.FullHouse,
                      ScoreType.SmallStraight,
                      ScoreType.LargeStraight,
                      ScoreType.Yahtzee,
                      ScoreType.Bonus
                      ]

    def __init__(self):
        self.dices = []
        self.score_map = {}
        self.reroll_times = 0
        self.score = 0

    @property
    def get_available_actions(self):
        return None

    def get_available_score_type(self):
        types = YahtzeeStatus.ALL_SCORE_TYPE

        return [x for x in types if self.get_score_by_type(x) is None]

    def get_score_by_type(self, score_type):
        if score_type in self.status_map:
            return self.score_map[score_type]

    @property
    def upper_sum(self):
        numerical = 0
        for x in [ScoreType.Ones,
                  ScoreType.Twos,
                  ScoreType.Threes,
                  ScoreType.Fours,
                  ScoreType.Fives,
                  ScoreType.Sixes]:
            if x in self.score_map:
                numerical += self.score_map[x]
        return numerical

    @property
    def status_code(self):
        state = 0
        # print(YahtzeeStatus.ALL_SCORE_TYPE, file=sys.stderr)

        for x in YahtzeeStatus.ALL_SCORE_TYPE[:-1]:
            if x in self.score_map:
                state += x

        self.dices.sort()
        dice_hash = 0
        for index, dice in enumerate(self.dices):
            dice_hash = dice_hash * 10 + dice

        numerical = self.upper_sum
        if numerical > 63:
            numerical = 63

        return state*100000*100*10 + dice_hash * 100 * 10 + numerical * 10 + self.reroll_times

    @classmethod
    def parse_status_code(cls, code):
        reroll_times = code % 10
        numerical = code // 10 % 100
        dice_hash = code // (10 * 100) % 100000
        state = code // (10 * 100 * 100000)
        return (state, dice_hash, numerical, reroll_times)

    @classmethod
    def compose_status_code(cls, state, dice_hash, numerical, reroll_times):
        return state*100000*100*10 + dice_hash * 100 * 10 + numerical * 10 + reroll_times

    @classmethod
    def compose_dices(cls, dices):
        dices = list(dices)
        dices.sort()
        dice_hash = 0
        for index, dice in enumerate(dices):
            dice_hash = dice_hash * 10 + dice
        return dice_hash

    _dices_hashmap = {}

    @classmethod
    def parse_dices(cls, dice_hash):
        if dice_hash in cls._dices_hashmap:
            return cls._dices_hashmap[dice_hash]

        dices = []
        t_dice_hash = dice_hash
        while t_dice_hash > 0:
            dices.append(t_dice_hash % 10)
            t_dice_hash //= 10
        dices.sort()
        cls._dices_hashmap[dice_hash] = dices
        return dices

    def copy(self):
        cloned = YahtzeeStatus()
        cloned.dices = self.dices.copy()
        cloned.score_map = self.score_map.copy()
        cloned.reroll_times = self.reroll_times
        cloned.turn = self.turn
        cloned.score = self.score
        return cloned

File: midterm/c++ search/YahtzeeStatus.cpp
#include "YahtzeeStatus.hpp"

#include <algorithm>
int YahtzeeStatus::getUpperSum() const
{
	static const auto list = std::vector<ScoreType>{
	ScoreType::Ones,
	ScoreType::Twos ,
	ScoreType::Threes ,
	ScoreType::Fours ,
	ScoreType::Fives,
	ScoreType::Sixes 
	};

	int score = 0;
	for (auto iter : list)
	{
		score += getScoreByType(iter);
	}
	return score;
}

int YahtzeeStatus::getLowerSum() const
{
	static const auto list = std::vector<ScoreType>{
	ScoreType::Chance,
	ScoreType::ThreeOfAKind ,
	ScoreType::FourOfAKind,
	ScoreType::FullHouse,
	ScoreType::SmallStraight,
	ScoreType::LargeStraight ,
	ScoreType::Yahtzee
	};

	int score = 0;
	for (auto iter : list)
	{
		score += getScoreByType(iter);
	}
	return score;
}

int YahtzeeStatus::getScoreByType(ScoreType type) const
{
	if (scoreMap.find(type) != scoreMap.end())
	{
		return scoreMap.at(type);
	}
	return 0;
}

bool YahtzeeStatus::end() const
{
	for (int i = 1; i < 13; i++)
	{
		ScoreType type = (ScoreType)(1 << i);
		if (scoreMap.find(type) == scoreMap.end())
		{
			return false;
		}
	}
	return true;
}

hash_t YahtzeeStatus::getStatusHash() const
{
	hash_t state = 0;

	for (auto pair : scoreMap)
	{
		if (pair.first != ScoreType::Bonus)
		{
			state |= (hash_t) pair.first;
		}
	}

	compressed_dicehash_t dicehash = composeDices(dices);

	int8_t upperSum = (int8_t)getUpperSum();
	if (upperSum > 63)
	{
		upperSum = 63;
	}
	return YahtzeeStatusCode{ state, dicehash, upperSum, (int8_t)reroll_times }.getHash();
}

int YahtzeeStatus::getDecDice(const Dices& dices)
{
	int res = 0;
	for (int i = 0; i < dices.size(); i++)
	{
		res = res * 10 + dices[i];
	}
	return res;
}

YahtzeeStatusCode YahtzeeStatus::parseStatusCode(hash_t code)
{
	return YahtzeeStatusCode::ParseHashCode(code);
}


compressed_dicehash_t YahtzeeStatus::composeDices(const Dices& dices)
{
	dicehash_t hash = 0;

	for (int dice : dices)
	{
		hash = (hash << 3) | dice;
	}

	return dicemap[hash];
}

compressed_dicehash_t YahtzeeStatus::composeDicehash(hash_t hash)
{
	return dicemap[hash];
}

Dices YahtzeeStatus::parseDices(compressed_dicehash_t compressedDiceHash)
{
	dicehash_t dicehash = reverse_dicemap[compressedDiceHash];
	Dices dices;
	bool valid = true;
	for (int range = 0; range < 5; range++)
	{
		int d = (dicehash >> (range * 3)) & 0x7;
		dices.push_back(d);
	}

	return dices;
}

void YahtzeeStatus::init()
{
	compressed_dicehash_t index = 0;
	for (dicehash_t dicehash = 0; dicehash < (1 << 15); dicehash++)
	{
		Dices dices;
		bool valid = true;
		for (int range = 0; range < 5; range++)
		{
			int d = (dicehash >> (range * 3)) & 0x7;
			dices.push_back(d);
			if (d == 7)
			{
				valid = false;
				break;
			}
		}

		if (!valid)
		{
			continue;
		}

		sort(dices.begin(), dices.end());

		dicehash_t sortedDicehash = 0;
		for (int range = 0; range < 5; range++)
		{
			sortedDicehash = (sortedDicehash << 3) | dices[range];
		}

		if (dicemap.find(sortedDicehash) != dicemap.end())
		{
			dicemap[dicehash] = dicemap[sortedDicehash];
		}
		else
		{
			dicemap[sortedDicehash] = index;
			dicemap[dicehash] = index;
			reverse_dicemap[index] = sortedDicehash;
			index++;
		}
	}
}

YahtzeeStatus YahtzeeStatus::getInitialState()
{
	YahtzeeStatus state;
	state.reroll_times = 2;
	return state;
}

std::vector<ScoreType> YahtzeeStatus::AllScoreType{
	ScoreType::Ones,
	ScoreType::Twos,
	ScoreType::Threes,
	ScoreType::Fours,
	ScoreType::Fives,
	ScoreType::Sixes,
	ScoreType::Chance,
	ScoreType::FourOfAKind,
	ScoreType::FullHouse,
	ScoreType::SmallStraight,
	ScoreType::LargeStraight,
	ScoreType::Yahtzee,
	ScoreType::Bonus
};

YahtzeeStatusCode YahtzeeStatusCode::ParseHashCode(hash_t code)
{
	YahtzeeStatusCode status;
	status.rerollTimes = code & 3;
	status.upperSum = (code >> 2) & 63;
	status.diceHash = (code >> 6 >> 2) & 0x1ff;
	status.state = (code >> 9 >> 6 >> 2);
	return status;
}

hash_t YahtzeeStatusCode::getHash()
{
	return state << 9 << 6 << 2 // 13 bits
		| diceHash << 6 << 2 // 9 bits
		| upperSum << 2 // 6 bits
		| rerollTimes; //2 bits
}

std::map<dicehash_t, compressed_dicehash_t> YahtzeeStatus::dicemap;
std::map<compressed_dicehash_t, dicehash_t> YahtzeeStatus::reverse_dicemap;
